### PR TITLE
subscription: Prevent Subscription spoke being mandatory with DVD source set

### PR DESCRIPTION
**Pull request by M4rtinK**
Looks like there is a chance for the Subscription spoke to be set mandatory if payload initialization takes too long and the Subscription spoke gets initialized before payload is done.

The first commit make the Subscription spoke initialization to wait on the payload thread to finish.

The second commit enhances this by moving the waiting to a separate thread, preventing it from potentially
blocking the GUI startup for a noticeable amount of time.

~**NOTE: testing in progress.**~
Works as expected with simulated CDROM detection delay.

### CODE REVIEW

File: pyanaconda/ui/gui/spokes/subscription.py [hunk 1]
```diff
@@ -634,8 +635,19 @@ def initialize(self):
...
+
+    def _initialize(self):
         # wait for subscription thread to finish (if any)
         threadMgr.wait(THREAD_SUBSCRIPTION)
+        # also wait for the payload thread, which migh still be processing
+        # a CDROM source, to avoid the Subscription being mandatory by mistake
+        # due to CDN still being default at the time of evaulation
+        threadMgr.wait(THREAD_PAYLOAD)
 
         # update overall state
```
Comment (poncovka) [inline pyanaconda/ui/gui/spokes/subscription.py:652]: I think that the methods bellow that update UI should have the `async_action_wait` decorator.

File: pyanaconda/ui/gui/spokes/subscription.py [hunk 2]
```diff
@@ -634,8 +635,19 @@ def initialize(self):
         # setup spoke state based on data from the Subscription DBus module
         self._update_spoke_state()
 
+        # start the rest of spoke initialization which might take some time
+        # (mainly due to waiting for various initialization threads to finish)
+        # in a separate thread
+        threadMgr.add(AnacondaThread(name=constants.THREAD_SUBSCRIPTION_SPOKE_INIT,
```
Comment (VladimirSlavik) [inline pyanaconda/ui/gui/spokes/subscription.py:641]: Isn't `constants.` redundant here? You refer to all the other threads directly by constant name...

**Comment by poncovka** [comment]
Jenkins, test this please.

File: pyanaconda/ui/gui/spokes/subscription.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/subscription.py]: Indeed, this was a copy paste typo & has been fixed.

File: pyanaconda/ui/gui/spokes/subscription.py (hunk 1 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/subscription.py:660]: Very good point! They are now running in a thread, which indeed could make GTK go boom.

File: pyanaconda/ui/gui/spokes/subscription.py [hunk 3]
```diff
@@ -126,8 +127,16 @@ def __init__(self, *args):
...
     def ready(self):
-        """The subscription spoke is always ready."""
-        return True
+        """The subscription spoke is ready once its initialization thread finishes.
+
+        We do this to avoid the Subscription spoke being set mandatory in cases
+        where the current installation source is the CDN, but payload refres is still
+        running and it might change to CDROM later one. We achieve this by waiting
+        for tha payload refresh thread to finish in the Subscription spoke initialization
+        thread.
```
Comment (VladimirSlavik) [inline pyanaconda/ui/gui/spokes/subscription.py:136]: refres -> refresh ; tha - > the

File: pyanaconda/ui/gui/spokes/subscription.py [hunk 4]
```diff
@@ -634,8 +643,19 @@ def initialize(self):
...
+    def _initialize(self):
         # wait for subscription thread to finish (if any)
         threadMgr.wait(THREAD_SUBSCRIPTION)
+        # also wait for the payload thread, which migh still be processing
+        # a CDROM source, to avoid the Subscription being mandatory by mistake
+        # due to CDN still being default at the time of evaulation
+        threadMgr.wait(THREAD_PAYLOAD)
 
         # update overall state
         self._update_registration_state()
```
Comment (VladimirSlavik) [inline pyanaconda/ui/gui/spokes/subscription.py:661]: Why do this twice?

File: pyanaconda/ui/gui/spokes/subscription.py (hunk 1 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/subscription.py:660]: Added the decorates, all seems fine after quick testing. (I have not hit a crash or freeze even before without the decorators, but that could have been luck as most of these GTK threading issues manifest as race conditions.

File: pyanaconda/ui/gui/spokes/subscription.py (hunk 3 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/subscription.py:136]: Fixed, thanks! :)

File: pyanaconda/ui/gui/spokes/subscription.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/subscription.py:661]: I'm not sure I understand, these are different methods:

```
        self._update_registration_state()
        self._update_subscription_state()
```

Or do you mean something else ?

Comment (VladimirSlavik) [reply pyanaconda/ui/gui/spokes/subscription.py:661]: Oh, wow, they are :O never mind then...

**Comment by M4rtinK** [comment]
jenkins, test this please

**Comment by M4rtinK** [comment]
Port to Fedora: #2807

**Pull request merged**